umbra/umbra_32.c

```diff
@@ -1049,3 +1049,53 @@ umbra_handle_fault(void *drcontext, byte *target, dr_mcontext_t *raw_mc,
 {
     return false;
 }
+
+drmf_status_t
+umbra_clear_redundant_blocks(umbra_map_t *map, uint *count)
+{
+    byte *shadow_data;
+    uint i, j;
+
+    if (map == NULL)
+        return DRMF_ERROR_INVALID_PARAMETER;
+
+    if (count != NULL)
+        *count = 0;
+
+    /* Umbra needs to have create-on-touch optimization enabled. */
+    if (!TEST(UMBRA_MAP_CREATE_SHADOW_ON_TOUCH, map->options.flags)){
+        return DRMF_ERROR_INVALID_PARAMETER;
+    }
+
+    umbra_map_lock(map);
+    for (i = 0; i < SHADOW_TABLE_ENTRIES; i++) {
+        shadow_data = shadow_table_get_block(map, i);
+        /* Redundant blocks must be "normal". */
+        if (shadow_table_is_in_normal_block(map, shadow_data)) {
```

Comment:
suggest: invert and continue to avoid indentation below

Reply:
Done.